lte/cloud/go/services/had/servicers/servicer.go

```diff
@@ -0,0 +1,263 @@
+	}
+
+	// All gateway pool records must have the same capacity, so use the first
+	// entry
+	callingRelativeCapacity := cellularCfg.Pooling[0].MmeRelativeCapacity
+	gwIDsToEnbs := map[string][]string{}
+	for _, record := range cellularCfg.Pooling {
+		gwIDsToEnbsInPool, err := s.getPrimaryGatewaysToEnodebs(secondaryGw.GetNetworkId(), string(record.GatewayPoolID), callingRelativeCapacity)
+		if err != nil {
+			return &protos.GetEnodebOffloadStateResponse{}, err
+		}
+		// Since a gateway can be in multiple pools, it is possible
+		// there could be key collisions. Since the ENBs values will be the
+		// same each time, this is okay.
+		for k, v := range gwIDsToEnbsInPool {
+			gwIDsToEnbs[k] = v
+		}
+	}
+	glog.V(2).Infof("Found the following primary gatewayIDs to ENB SNs: %v", gwIDsToEnbs)
+
+	enbSNsToOffloadState := map[uint32]protos.GetEnodebOffloadStateResponse_EnodebOffloadState{}
+	for primaryGwID, enbs := range gwIDsToEnbs {
+		isCheckinValid, err := s.isGatewayCheckinValid(secondaryGw.NetworkId, primaryGwID)
+		if err != nil {
+			glog.Error(err)
+			continue
+		} else if !isCheckinValid {
+			continue
+		}
+		for _, enb := range enbs {
+			offloadState, err := s.getOffloadStateForEnb(secondaryGw.NetworkId, primaryGwID, enb)
+			if err != nil {
+				glog.Error(err)
+				continue
+			}
+			enbID, err := s.getEnodebID(secondaryGw.NetworkId, secondaryGw.LogicalId, enb)
+			if err != nil {
+				glog.Error(err)
+				continue
+			}
+			enbSNsToOffloadState[enbID] = offloadState
+		}
+	}
+	return &protos.GetEnodebOffloadStateResponse{
+		EnodebOffloadStates: enbSNsToOffloadState,
+	}, nil
+}
+
+func (s *HAServicer) getPrimaryGatewaysToEnodebs(networkID string, gatewayPoolID string, callingRelativeCapacity uint32) (map[string][]string, error) {
+	poolEnt, err := configurator.LoadEntity(
+		networkID, lte.CellularGatewayPoolEntityType, gatewayPoolID,
+		configurator.EntityLoadCriteria{LoadAssocsFromThis: true},
+		lte_models.EntitySerdes,
+	)
+	if err != nil {
+		return map[string][]string{}, err
+	}
+	primaryGatewaysToEnbs := map[string][]string{}
+	for _, gw := range poolEnt.Associations.Filter(lte.CellularGatewayEntityType) {
+		ent, err := configurator.LoadEntity(networkID, lte.CellularGatewayEntityType, gw.Key,
+			configurator.EntityLoadCriteria{LoadConfig: true, LoadAssocsFromThis: true, LoadAssocsToThis: true},
+			lte_models.EntitySerdes,
+		)
+		if err != nil {
+			return map[string][]string{}, err
+		}
+		cellularCfg, ok := ent.Config.(*lte_models.GatewayCellularConfigs)
+		if !ok {
+			return map[string][]string{}, fmt.Errorf("could not convert stored config to type GatewayCellularConfigs for gw %s", gw.Key)
+		}
+		if cellularCfg.Pooling == nil || len(cellularCfg.Pooling) == 0 {
+			return map[string][]string{}, fmt.Errorf("gateway '%s' is not configured in a gateway pool", gw.Key)
+		}
+		// primary gateways are those with strictly higher relative capacity
```

Comment:
Do we enforce this invariant at the API layer?

Reply:
Yeah we do. It's enforced in the ValidateModel method for the Pooling config.